### PR TITLE
fix: Allow separating by spaces and/or commas

### LINKED ISSUE #4564: MatrixTransform.Matrix without comma separators generates incorrect C# code
## Current behavior

Errors are thrown if commas are not added to separate the individual values in the `MatrixTransform.Matrix` value, even though that is valid XAML syntax.

## Expected behavior

Commas should be inserted between the values so that the resulting string can be inserted into the `MatrixTransform` constructor correctly.

## How to reproduce it

Specify a `MatrixTransform` somewhere in your code as follows:

```xml
<MatrixTransform Matrix="1.1048446 0 0 1.1959561 -67.035883 -96.098854"/>
```
The following invalid C# code will be generated, breaking the compilation:
```cs
new Matrix(1.1048446 0 0 1.1959561 -67.035883 -96.098854);
```

## Environment

Nuget Package:

- [X] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other:

Nuget Package Version(s):

`Uno.UI.WebAssembly`: v3.3.0

Affected platform(s):

- [ ] iOS
- [ ] Android
- [X] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

IDE:

- [ ] Visual Studio 2017 (version: )
- [X] Visual Studio 2019 (version: 16.8)
- [ ] Visual Studio for Mac (version: )
- [ ] Rider Windows (version: )
- [ ] Rider macOS (version: )
- [ ] Visual Studio Code (version: )

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #4564

## PR Type

- Bugfix

## What is the current behavior?

The following isn't allowed and generates invalid code.

```xml
<MatrixTransform Matrix="1.1048446 0 0 1.1959561 -67.035883 -96.098854"/>
```

## What is the new behavior?

It's now allowed.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by MartinZikmund — APPROVED** [review]

**Review by davidjohnoliver — COMMENTED** [review]
Just needs some tests :)

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 1]
```diff
@@ -997,6 +995,9 @@ private static bool FastStringToTypeConvert(Type outputType, string input, ref o
 				return false;
 			}
 		}
+
+		static string[] Split(string value)
+			=> Regex.Split(value, @"(\s*?,\s*|\s+)");
```
Comment (dr1rrb) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:1000]:
```suggestion
			=> Regex.Split(value, @"(\s*,\s*|\s+)");
```

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 2]
```diff
@@ -4442,6 +4442,9 @@ string Inner()
 				}
 
 				throw new Exception("Unable to convert {0} for {1} with type {2}".InvariantCultureFormat(memberValue, memberName, propertyType));
+
+				static string? splitAndJoin(string? value)
+					=> value == null ? null : string.Join(", ", Regex.Split(value, @"(\s*?,\s*|\s+)"));
```
Comment (dr1rrb) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:4447]: Am I wrong thinking that a `Regex.Replace` would be sufficient here? Also `*?` in the regex looks redundant to me :)

```suggestion
					=> value == null ? null : Regex.Replace(value, @"(\s*,\s*|\s+)", ", "));
```

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 3]
```diff
@@ -4442,6 +4442,9 @@ string Inner()
 				}
 
 				throw new Exception("Unable to convert {0} for {1} with type {2}".InvariantCultureFormat(memberValue, memberName, propertyType));
+
+				static string? splitAndJoin(string? value)
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:4446]: The debate is not yet completely settled, but local functions should use pascal casing. We'll need to add that to the coding guidelines.

```suggestion
				static string? SplitAndJoin(string? value)
```

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 4]
```diff
@@ -997,6 +995,9 @@ private static bool FastStringToTypeConvert(Type outputType, string input, ref o
 				return false;
 			}
 		}
+
+		static string[] Split(string value)
+			=> Regex.Split(value, @"(\s*,\s*|\s+)");
```
Comment (jeromelaban) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:1000]: We should create the Regex only once so there's no additional parsing donc on each split:

```suggestion
      static readonly Regex _matrixSplit = new Regex(@"(\s*,\s*|\s+)");
        
		static string[] Split(string value)
			=> _matrixSplit.Split(value);
```

**Review by jeromelaban — APPROVED** [review]

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs (hunk 3 above)
Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:4446]: Will fix that.

I'd suggest enforcing coding-style via globalconfig or editorconfig.

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:1000]: I'm not using the variable specifically for matrix. Will rename it

**Comment by jeromelaban** [comment]
Looks like the formatting change broke a XamlReader test: https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=31438&view=ms.vss-test-web.build-test-results-tab&runId=1531484&paneView=debug&resultId=100047

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 5]
```diff
@@ -154,6 +154,7 @@ internal partial class XamlFileGenerator
 		/// If set, code generated from XAML will be annotated with the source method and line # in this file, for easier debugging.
 		/// </summary>
 		private readonly bool _shouldAnnotateGeneratedXaml;
+		private static readonly Regex splitRegex = new Regex(@"(\s*,\s*|\s+)");
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:157]:
```suggestion
		private static readonly Regex splitRegex = new Regex(@"\s*,\s*|\s+");
```

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 6]
```diff
@@ -997,6 +995,11 @@ private static bool FastStringToTypeConvert(Type outputType, string input, ref o
 				return false;
 			}
 		}
+
+		private static readonly Regex _splitRegex = new Regex(@"(\s*,\s*|\s+)");
```
Comment (Youssef1313) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:999]:
```suggestion
		private static readonly Regex _splitRegex = new Regex(@"\s*,\s*|\s+");
```

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**